### DISCUSSION TITLE
Rspec testing and Rodauth-rails - signing in a user

### DISCUSSION THREAD

**Discussion by wdperson** [janko/rodauth-rails#101] [Q&A]
I am trying to write some tests using Rspec with Rodauth-Rails.  

Here is what I have:

In a rspec helper: 
```
  def sign_in(user)
    visit "/login"
    fill_in "login", with: user.email
    fill_in "password", with: user.password
    click_on "Login"
  end
```

rspec controller test:
```
  let(:user) { create(:user) }

  before(:each) do
    sign_in(user)
  end

  describe 'GET index' do
    it "responds successfully" do
      get :show
      expect(response).to be_success
    end
  end
```

So, all I am doing is trying to sign in a user since this is blocked by authentication.  However, using capybara and filling in the fields doesn't work and throws the following error:
```
     KeyError:
       key not found: "rodauth"
     # /usr/local/bundle/gems/rodauth-rails-0.17.1/lib/rodauth/rails/controller_methods.rb:12:in `fetch'
     # /usr/local/bundle/gems/rodauth-rails-0.17.1/lib/rodauth/rails/controller_methods.rb:12:in `rodauth'
```

Any ideas on what the issue is here?

thanks
--Mike Riley

**Comment by janko** [comment]
The issue is that controllers rely on `env["rodauth"]` to be set (this is what the `#rodauth` controller method reads), which is set in the Rodauth middleware, but controller tests don't execute the middleware stack. I will see if I can come up with similar test helpers as Devise has.

**Reply by wdperson** [reply]
thanks @janko.  I appreciate it.  I am more than happy to help test should you need it.

**Reply by janko** [reply]
I just opened a pull request for this – https://github.com/janko/rodauth-rails/pull/102. I would appreciate it if you could test it out 🙂
